velox/exec/tests/RowContainerTest.cpp

```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+                                       const std::pair<int, char*>& r) {
+          return rowContainer->compare(
+                     l.second,
+                     rowContainer->columnAt(0),
+                     decoded,
+                     r.first,
+                     {nullsFirst, ascending}) < 0;
+        };
+
+    if (ascending && nullsFirst) {
+      sortDecodedVectorAndRows = [&](const std::pair<int, char*>& l,
+                                     const std::pair<int, char*>& r) {
+        return rowContainer->compare(
+                   l.second, rowContainer->columnAt(0), decoded, r.first) < 0;
+      };
     }
 
-    // Verify compareRows method with row as input.
+    // Verify compare method with row and decoded vector as input - using
+    // sortDecodedVectorAndRows
     {
       std::sort(
-          indexedRows.begin(),
-          indexedRows.end(),
-          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
-            return rowContainer->compareRows(
-                       l.second, r.second, {{nullsFirst, ascending}}) < 0;
-          });
-      std::vector<std::optional<T>> sorted;
-      for (const auto& irow : indexedRows) {
-        if (decoded.isNullAt(irow.first)) {
-          sorted.push_back(std::nullopt);
-        } else {
-          sorted.push_back(decoded.valueAt<T>(irow.first));
-        }
-      }
-      VectorPtr result = vectorMaker.template flatVectorNullable<T>(sorted);
+          indexedRows.begin(), indexedRows.end(), sortDecodedVectorAndRows);
+      auto result = copyValues(indexedRows, values);
       assertEqualVectors(expected, result);
     }
-    // Verify compareRows method with default compare flags.
-    if (ascending && nullsFirst) {
-      std::sort(
-          indexedRows.begin(),
-          indexedRows.end(),
-          [&](const std::pair<int, char*>& l, const std::pair<int, char*>& r) {
-            return rowContainer->compareRows(l.second, r.second) < 0;
-          });
-      std::vector<std::optional<T>> sorted;
-      for (const auto& irow : indexedRows) {
-        if (decoded.isNullAt(irow.first)) {
-          sorted.push_back(std::nullopt);
-        } else {
-          sorted.push_back(decoded.valueAt<T>(irow.first));
+  }
+
+  template <typename T>
+  void
+  testComparePrimitiveFloats(TypePtr type, bool ascending, bool nullsFirst) {
+    // NOTE: set 'isJoinBuild' to true to enable nullable sort key in test.
+    auto lowest = std::numeric_limits<T>::lowest();
+    auto min = std::numeric_limits<T>::min();
+    auto max = std::numeric_limits<T>::max();
+    auto nan = std::numeric_limits<T>::quiet_NaN();
+    auto inf = std::numeric_limits<T>::infinity();
+
+    VectorPtr values = makeNullableFlatVector<T>(
+        {std::nullopt, max, inf, 0.0, nan, lowest, min});
+
+    std::vector<std::optional<T>> expectedOrder = {
+        std::nullopt, lowest, 0.0, min, max, inf, nan};
+    changeSortingOrder<T>(expectedOrder, ascending, nullsFirst);
```

Comment:
Maybe rearrange the code to avoid re-creating values and expectedOrder for each combination of compare flags. E.g. remove the flags from the signature and add a loop here.